docs/adr/0003-threadsafe-interfaces.md

```diff
@@ -0,0 +1,196 @@
+
+### Add a `[Threadsafe]` attribute to UDL `interface` definitions.
+
+We would advise Nimbus SDK to update their UDL definition for `NimbusClient` like so:
+
+```idl
+[Threadsafe]
+interface NimbusClient {
+  // existing method definitions remain unchanged
+}
+```
+
+The [`uniffi_bindgen::interface::Object`](https://github.com/mozilla/uniffi-rs/blob/803bb3d79daa8ea088fb2d8f05c08ada09821986/uniffi_bindgen/src/interface/mod.rs#L722) struct
+would grow a corresponding boolean `threadsafe` field and corresponding public accessor.
+When [building an instance of this struct from the UDL](https://github.com/mozilla/uniffi-rs/blob/803bb3d79daa8ea088fb2d8f05c08ada09821986/uniffi_bindgen/src/interface/mod.rs#L786),
+we would inspect the list of attributes to look for one named "Threadsafe" and set the field
+if present. The way that we [handle the `[ByRef]` annotation on method arugments](https://github.com/mozilla/uniffi-rs/blob/803bb3d79daa8ea088fb2d8f05c08ada09821986/uniffi_bindgen/src/interface/mod.rs#L658)
+will likely serve as a good example to follow.
+
+### Implement a less-locking HandleMap variant
+
+The `ffi_support` crate provides a basic non-locking [`HandleMap<T>`](https://docs.rs/ffi-support/0.4.0/ffi_support/handle_map/struct.HandleMap.html) struct,
+and it implements `ConcurrentHandleMap` as a thin wrapper around a `RwLock<HandleMap<Mutex<T>>>`. We can make our own variant that removes the inner mutex, as a thin wrapper around a`RwLock<HandleMap<T>>`. (The outer `RwLock` is still needed in order to guard mutations of the handlemap itself, and we must be careful to only hold it for short periods).
+
+Bikeshed name: `NonLockingHandleMap`, implemented under the [`uniffi::ffi` module](https://github.com/mozilla/uniffi-rs/tree/main/uniffi/src/ffi). It needs to live in the `uniffi` crate so that it can be used by the generated Rust scaffolding at runtime.
+
+We don't need to support the entire `ConcurrentHandleMap` interface, only:
+* `insert_with_output`
+* `insert_with_result`
+* `call_with_output`
+* `call_with_result`
+* `delete_u64`
+
+The `NonLockingHandleMap<T>` struct should be capable of handing out `&T` references but should never hand
+out a `&mut T`. It should also require that `T: Send + Sync`.
+
+### Use `NonLockingHandleMap` in the Rust scaffolding for `[Threadsafe]` instances
+
+In [ObjectTemplate.rs](https://github.com/mozilla/uniffi-rs/blob/main/uniffi_bindgen/src/templates/ObjectTemplate.rs), check the `threadsafe` property of the object definition. If it's true, use `NonLockingHandleMap` instead of `ConcurrentHandleMap` in the [lazy static that declares the handlemap for that interface](https://github.com/mozilla/uniffi-rs/blob/803bb3d79daa8ea088fb2d8f05c08ada09821986/uniffi_bindgen/src/templates/ObjectTemplate.rs#L11).
+
+By ensuring that `NonLockingHandleMap` and `ConcurrentHandleMap` expose a similar API, this will hopefully
+be a fairly minimal if-then-else that just chooses the correct name of the struct to use.
```

Comment:
I expect we're going to have to delve into the method bodies themselves, to add or remove `mut` annotations in various places.

Reply:
Yes, I think you're right; we may even find we want to define a small `HandleMap` trait of our own and adapt `ConcurrentHandleMap` to it, rather than tying ourselves to to the API it already offers.